# COMS W4705 Spring 24
## Homework 4 - Semantic Role Labelling with BERT

The goal of this assignment is to train and evaluate a PropBank-style semantic role labeling (SRL) system. Following (Collobert et al. 2011) and others, we will treat this problem as a sequence-labeling task. For each input token, the system will predict a B-I-O tag, as illustrated in the following example: 

|The|judge|scheduled|to|preside|over|his|trial|was|removed|from|the|case|today|.|             
|---|-----|---------|--|-------|----|---|-----|---|-------|----|---|----|-----|-|             
|B-ARG1|I-ARG1|B-V|B-ARG2|I-ARG2|I-ARG2|I-ARG2|I-ARG2|O|O|O|O|O|O|O|
|||schedule.01|||||||||||||

Note that the same sentence may have multiple annotations for different predicates

|The|judge|scheduled|to|preside|over|his|trial|was|removed|from|the|case|today|.|             
|---|-----|---------|--|-------|----|---|-----|---|-------|----|---|----|-----|-|             
|B-ARG1|I-ARG1|I-ARG1|I-ARG1|I-ARG1|I-ARG1|I-ARG1|I-ARG1|O|B-V|B-ARG2|I-ARG2|I-ARG2|B-ARGM-TMP|O|
||||||||||remove.01||||||

and not all predicates need to be verbs

|The|judge|scheduled|to|preside|over|his|trial|was|removed|from|the|case|today|.|             
|---|-----|---------|--|-------|----|---|-----|---|-------|----|---|----|-----|-|    
|O|O|O|O|O|O|B-ARG1|B-V|O|O|O|O|O|O|O|
||||||||try.02||||||||

The SRL system will be implemented in [PyTorch](https://pytorch.org/). We will use BERT (in the implementation provided by the [Huggingface transformers](https://huggingface.co/docs/transformers/index) library) to compute contextualized token representations and a custom classification head to predict semantic roles. We will fine-tune the pretrained BERT model on the SRL task. 


### Overview of the Approach

The model we will train is pretty straightforward. Essentially, we will just encode the sentence with BERT, then take the contextualized embedding for each token and feed it into a classifier to predict the corresponding tag. 

Because we are only working on argument identification and labeling (not predicate identification), it is essentially that we tell the model where the predicate is. This can be accomplished in various ways. The approach we will choose here repurposes Bert's *segment embeddings*. 

Recall that BERT is trained on two input sentences, seperated by [SEP], and on a next-sentence-prediction objective (in addition to the masked LM objective). To help BERT comprehend which sentence a given token belongs to, the original BERT uses a segment embedding, using A for the first sentene, and B for the second sentence 2. 
Because we are labeling only a single sentence at a time, we can use the segment embeddings to indicate the predicate position instead: The predicate is labeled as segment B (1) and all other tokens will be labeled as segment A (0). 

<img src="https://github.com/daniel-bauer/4705-f23-hw5/blob/main/bert_srl_model.png?raw=true" width=400px>

## Setup: GCP, Jupyter, PyTorch, GPU

To make sure that PyTorch is available and can use the GPU,run the following cell which should return True. If it doesn't, make sure the GPU drivers and CUDA are installed correctly. 

GPU support is required for this assignment -- you will not be able to fine-tune BERT on a CPU. 

In [1]:
import torch
torch.cuda.is_available() 

True

## Dataset: Ontonotes 5.0 English SRL annotations

We will work with the English part of the [Ontonotes 5.0](https://catalog.ldc.upenn.edu/LDC2013T19) data. This is an extension of PropBank, using the same type of annotation. Ontonotes contains annotations other than predicate/argument structures, but we will use the PropBank style SRL annotations only. *Important*: This data set is provided to you for use in COMS 4705 only! Columbia is a subscriber to LDC and is allowed to use the data for educational purposes. However, you may not use the dataset in projects unrelated to Columbia teaching or research.

If you haven't done so already, you can download the data here: 


In [2]:
! wget https://storage.googleapis.com/4705-bert-srl-data/ontonotes_srl.zip    

--2024-04-23 13:32:26--  https://storage.googleapis.com/4705-bert-srl-data/ontonotes_srl.zip
Resolving storage.googleapis.com (storage.googleapis.com)... 172.217.31.27, 142.251.130.27, 142.251.222.219, ...
Connecting to storage.googleapis.com (storage.googleapis.com)|172.217.31.27|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12369688 (12M) [application/zip]
Saving to: ‘ontonotes_srl.zip’

ontonotes_srl.zip   100%[===================>]  11.80M  7.14MB/s    in 1.7s    

2024-04-23 13:32:29 (7.14 MB/s) - ‘ontonotes_srl.zip’ saved [12369688/12369688]



In [3]:
! unzip ontonotes_srl.zip    

Archive:  ontonotes_srl.zip
  inflating: propbank_dev.tsv        
  inflating: propbank_test.tsv       
  inflating: propbank_train.tsv      
  inflating: role_list.txt           


The data has been pre-processed in the following format. There are three files: 

`propbank_dev.tsv`	`propbank_test.tsv`	`propbank_train.tsv`

Each of these files is in a tab-separated value format. A single predicate/argument structure annotation consists of four rows. For example 

```
ontonotes/bc/cnn/00/cnn_0000.152.1
The     judge   scheduled       to      preside over    his     trial   was     removed from    the     case    today   /.
                schedule.01
B-ARG1  I-ARG1  B-V     B-ARG2  I-ARG2  I-ARG2  I-ARG2  I-ARG2  O       O       O       O       O       O       O
```

* The first row is a unique identifier (1st annotation of the 152nd sentence in the file ontonotes/bc/cnn/00/cnn_0000).
* The second row contains the tokens of the sentence (tab-separated). 
* The third row contains the probank frame name for the predicate (empty field for all other tokens). 
* The fourth row contains the B-I-O tag for each token. 

The file `rolelist.txt` contains a list of propbank BIO labels in the dataset (i.e. possible output tokens). This list has been filtered to contain only roles that appeared more than 1000 times in the training data. 
We will load this list and create mappings from numeric ids to BIO tags and back.

In [93]:
role_to_id = {}
with open("role_list.txt",'r') as f:
    role_list = [x.strip() for x in f.readlines()]
    role_to_id = dict((role, index) for (index, role) in enumerate(role_list))
    role_to_id['[PAD]'] = -100
#     role_to_id['O'] = 0
    
    id_to_role = dict((index, role) for (role, index) in role_to_id.items())
    

Note that we are also mapping the '[PAD]' token to the value -100. This allows the loss function to ignore these tokens during training.

In [94]:
# Printing the top 10 key-value pairs
print("Top 10 key-value pairs in role_to_id dictionary:")
for key, value in list(role_to_id.items()):
    print(f"{key}: {value}")

Top 10 key-value pairs in role_to_id dictionary:
[PAD]: -100
[CLS]: 1
[SEP]: 2
[prd]: 3
O: 4
B-ARG0: 5
B-ARG1: 6
B-ARG1-DSP: 7
B-ARG2: 8
B-ARG3: 9
B-ARG4: 10
B-ARGM-ADJ: 11
B-ARGM-ADV: 12
B-ARGM-CAU: 13
B-ARGM-DIR: 14
B-ARGM-DIS: 15
B-ARGM-EXT: 16
B-ARGM-GOL: 17
B-ARGM-LOC: 18
B-ARGM-MNR: 19
B-ARGM-MOD: 20
B-ARGM-NEG: 21
B-ARGM-PRD: 22
B-ARGM-PRP: 23
B-ARGM-PRR: 24
B-ARGM-TMP: 25
B-ARGM-CXN: 26
B-ARGM-LVB: 27
B-V: 28
I-ARG0: 29
I-ARG1: 30
I-ARG1-DSP: 31
I-ARG2: 32
I-ARG3: 33
I-ARG4: 34
I-ARGM-ADJ: 35
I-ARGM-ADV: 36
I-ARGM-CAU: 37
I-ARGM-DIR: 38
I-ARGM-DIS: 39
I-ARGM-EXT: 40
I-ARGM-GOL: 41
I-ARGM-LOC: 42
I-ARGM-MNR: 43
I-ARGM-MOD: 44
I-ARGM-NEG: 45
I-ARGM-PRD: 46
I-ARGM-PRP: 47
I-ARGM-PRR: 48
I-ARGM-TMP: 49
I-ARGM-CXN: 50
I-ARGM-LVB: 51
I-V: 52


## Part 1 - Data Preparation

Before you can build the SRL model, you first need to preprocess the data. 


### 1.1 - Tokenization 

One challenge is that the pre-trained BERT model uses subword ("WordPiece") tokenization, but the Ontonotes data does not. Fortunately Huggingface transformers provides a tokenizer. 

In [421]:
from transformers import BertTokenizerFast 
tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased', do_lower_case=True)
tokenizer.tokenize("This is an unbelievably boring test sentence.")

['this',
 'is',
 'an',
 'un',
 '##bel',
 '##ie',
 '##va',
 '##bly',
 'boring',
 'test',
 'sentence',
 '.']

**TODO**: 
We need to be able to maintain the correct labels (B-I-O tags) for each of the subwords. 
Complete the following function that takes a list of tokens and a list of B-I-O labels of the same length as parameters, and returns a new token / label pair, as illustrated in the following example. 


```
>>> tokenize_with_labels("the fancyful penguin devoured yummy fish .".split(), "B-ARG0 I-ARG0 I-ARG0 B-V B-ARG1 I-ARG1 O".split(), tokenizer)
(['the',
  'fancy',
  '##ful',
  'penguin',
  'dev',
  '##oured',
  'yu',
  '##mmy',
  'fish',
  '.'],
 ['B-ARG0',
  'I-ARG0',
  'I-ARG0',
  'I-ARG0',
  'B-V',
  'I-V',
  'B-ARG1',
  'I-ARG1',
  'I-ARG1',
  'O'])

```

To approach this problem, iterate through each word/label pair in the sentence. Call the tokenizer on the word. This may result in one or more tokens. Create the correct number of labels to match the number of tokens. Take care to not generate multiple B- tokens. 


This approach is a bit slower than tokenizing the entire sentence, but is necessary to produce proper input tokenization for the pre-trained BERT model, and the matching target labels. 

In [422]:
def tokenize_with_labels(sentence, text_labels, tokenizer):
    """
    Word piece tokenization makes it difficult to match word labels
    back up with individual word pieces. 
    """

    tokenized_sentence = []
    labels = []
    
    for word, label in zip(sentence, text_labels):
        word_tokens = tokenizer.tokenize(word)
        
        tokenized_sentence.extend(word_tokens)
        
        if len(word_tokens) > 1:
            label_parts = label.split('-')
            if label_parts[0] == 'B':
                labels.extend([label] + ['I-' + '-'.join(label_parts[1:])] * (len(word_tokens) - 1))
            elif label_parts[0] == 'I':
                labels.extend(['I-' + '-'.join(label_parts[1:])] * len(word_tokens))
            else:
                labels.extend(['O'] * len(word_tokens))
        else:
            labels.append(label)

    return tokenized_sentence, labels

In [423]:
tokenize_with_labels("the fancyful penguin devoured yummy fish .".split(), "B-ARG0 I-ARG0 I-ARG0 B-V B-ARG1 I-ARG1 O".split(), tokenizer)

(['the',
  'fancy',
  '##ful',
  'penguin',
  'dev',
  '##oured',
  'yu',
  '##mmy',
  'fish',
  '.'],
 ['B-ARG0',
  'I-ARG0',
  'I-ARG0',
  'I-ARG0',
  'B-V',
  'I-V',
  'B-ARG1',
  'I-ARG1',
  'I-ARG1',
  'O'])

In [424]:
tokenize_with_labels("We respectfully invite	you to watch a special edition of Across China .".split(), "B-ARG0	B-ARGM-MNR	B-V	B-ARG1	B-ARG2	I-ARG2	I-ARG2	I-ARG2	I-ARG2	I-ARG2	I-ARG2	I-ARG2	O".split(), tokenizer)

(['we',
  'respectful',
  '##ly',
  'invite',
  'you',
  'to',
  'watch',
  'a',
  'special',
  'edition',
  'of',
  'across',
  'china',
  '.'],
 ['B-ARG0',
  'B-ARGM-MNR',
  'I-ARGM-MNR',
  'B-V',
  'B-ARG1',
  'B-ARG2',
  'I-ARG2',
  'I-ARG2',
  'I-ARG2',
  'I-ARG2',
  'I-ARG2',
  'I-ARG2',
  'I-ARG2',
  'O'])

### 1.2 Loading the Dataset

Next, we are creating a PyTorch [Dataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset) class. This class acts as a contained for the training, development, and testing data in memory. You should already be familiar with Datasets and Dataloaders from homework 3. 

1.2.1 **TODO**: Write the \_\_init\_\_(self, filename) method that reads in the data from a data file (specified by the filename).

For each annotation you start with  the tokens in the sentence, and the BIO tags. Then you need to create the following 

1. call the `tokenize_with_labels` function to tokenize the sentence.
2. Add the (token, label) pair to the self.items list. 

1.2.2 **TODO**: Write the \_\_len\_\_(self) method that returns the total number of items. 

1.2.3 **TODO**: Write the \_\_getitem\_\_(self, k) method that returns a single item in a format BERT will understand. 
* We need to process the sentence by adding "\[CLS\]" as the first token and "\[SEP\]" as the last token. The need to pad the token sequence to 128 tokens using the "\[PAD\]" symbol. This needs to happen both for the inputs (sentence token sequence) and outputs (BIO tag sequence).
* We need to create an *attention mask*, which is a sequence of 128 tokens indicating the actual input symbols (as a 1) and \[PAD\] symbols (as a 0).
* We need to create a *predicate indicator* mask, which is a sequence of 128 tokens with at most one 1, in the position of the "B-V" tag. All other entries should be 0. The model will use this information to understand where the predicate is located. 

* Finally, we need to convert the token and tag sequence into numeric indices. For the tokens, this can be done using the `tokenizer.convert_tokens_to_ids` method. For the tags, use the `role_to_id` dictionary. 
Each sequence must be a pytorch tensor of shape (1,128). You can convert a list of integer values like this `torch.tensor(token_ids, dtype=torch.long)`.

To keep everything organized, we will return a dictionary in the following format 

```
{'ids': token_tensor,
 'targets': tag_tensor,
 'mask': attention_mask_tensor,
 'pred': predicate_indicator_tensor}
```


(Hint: To debug these, read in the first annotation only / the first few annotations) 


In [464]:
from torch.utils.data import Dataset, DataLoader 

class SrlData(Dataset):
    
    def __init__(self, filename):

        super(SrlData, self).__init__()
        
        self.max_len = 128 
        
        self.tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased', do_lower_case=True)        
        
        self.items = []
        
        with open(filename, 'r') as file:
            lines = file.readlines()
            num_lines = len(lines)
            idx = 0
            
            while idx < num_lines:
                identifier = lines[idx].strip()  
                sentence = lines[idx + 1].strip().split()  
                _ = lines[idx + 2] 
                labels = lines[idx + 3].strip().split()  
                try:
                    tokenized_sentence, new_labels = tokenize_with_labels(sentence, labels, self.tokenizer)
                    self.items.append((tokenized_sentence, new_labels))
                    
                except ValueError as e:
                    print("Error:", e)
                idx += 4
                                                        
    def __len__(self):
        return len(self.items)
    
    def __getitem__(self, k):
        tokenized_sentence, labels = self.items[k]
        
        
        # Add [CLS] and [SEP] tokens
        tokenized_sentence = ["[CLS]"] + tokenized_sentence + ["[SEP]"]
        token_labels = ["[CLS]"] + labels + ["[SEP]"]
        
        # Check if the length exceeds max_len after adding special tokens
        if len(tokenized_sentence) > self.max_len:
#             print("original")
#             print(tokenized_sentence)
#             print(labels)
            tokenized_sentence = tokenized_sentence[:self.max_len - 1] + ["[SEP]"]  # Keep [SEP] at the end
            token_labels = token_labels[:self.max_len - 1] + ["[PAD]"]  # Padding for labels
#             print("Padding if len higher")
#             print(tokenized_sentence)
#             print(token_labels)
            
        # Pad sequences to max_length
        pad_length = self.max_len - len(tokenized_sentence)
        tokenized_sentence += ["[PAD]"] * pad_length
        token_labels += ["[PAD]"] * pad_length

        token_ids = tokenizer.convert_tokens_to_ids(tokenized_sentence)
        # Assuming token_labels is a list of role labels
        label_ids = [role_to_id.get(label, role_to_id['O']) for label in token_labels]

        attention_mask = [1 if token != "[PAD]" else 0 for token in tokenized_sentence]
        predicate_indicator = [1 if label == "B-V" else 0 for label in token_labels]
        
        token_ids = torch.tensor(token_ids, dtype=torch.long).unsqueeze(0)
        label_ids = torch.tensor(label_ids, dtype=torch.long).unsqueeze(0)
        attention_mask = torch.tensor(attention_mask, dtype=torch.long).unsqueeze(0)
        predicate_indicator = torch.tensor(predicate_indicator, dtype=torch.long).unsqueeze(0)

        return {
            'ids': token_ids,
            'targets': label_ids,
            'mask': attention_mask,
            'pred': predicate_indicator
        }
        

In [465]:
# Open the input file
with open('propbank_train.tsv', 'r') as input_file:
    # Read the first 10 lines
    lines = input_file.readlines()[:8000]

# Open the output file
with open('output.tsv', 'w') as output_file:
    # Write the lines to the output file
    output_file.writelines(lines)

In [466]:
# Reading the training data takes a while for the entire data because we preprocess all data offline
data = SrlData("propbank_train.tsv")

In [467]:
idx = 2  # Choose the index of the item you want to retrieve
item = data[idx]

# Print the retrieved item
print("Retrieved item:")
print(item)

Retrieved item:
{'ids': tensor([[  101,  3061,  4206,  2006, 13843, 18003,  3137,  2003,  1996,  6104,
          2000,  1996,  3634, 10435,  5805,  1012,   102,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0

## 2. Model Definition

In [468]:
from torch.nn import Module, Linear, CrossEntropyLoss
from transformers import BertModel

We will define the pyTorch model as a subclass of the [torch.nn.Module](https://pytorch.org/docs/stable/generated/torch.nn.Module.html) class. The code for the model is provided for you. It may help to take a look at the documentation to remind you of how Module works. Take a look at how the huggingface BERT model simply becomes another sub-module. 

In [469]:
class SrlModel(Module):
    
    def __init__(self):
        
        super(SrlModel, self).__init__()
        
        self.encoder = BertModel.from_pretrained("bert-base-uncased")
    
        # The following two lines would freeze the BERT parameters and allow us to train the classifier by itself.
        # We are fine-tuning the model, so you can leave this commented out!
        # for param in self.encoder.parameters():
        #    param.requires_grad = False
        
        # The linear classifier head, see model figure in the introduction. 
        self.classifier = Linear(768, len(role_to_id))
                
        
    def forward(self, input_ids, attn_mask, pred_indicator):
    
        # This defines the flow of data through the model 
    
        # Note the use of the "token type ids" which represents the segment encoding explained in the introduction. 
        # In our segment encoding, 1 indicates the predicate, and 0 indicates everything else. 
        bert_output =  self.encoder(input_ids=input_ids, attention_mask=attn_mask, token_type_ids=pred_indicator)

        enc_tokens = bert_output[0] # the result of encoding the input with BERT
        logits = self.classifier(enc_tokens) #feed into the classification layer to produce scores for each tag.
        
        # Note that we are only interested in the argmax for each token, so we do not have to normalize 
        # to a probability distribution using softmax. The CrossEntropyLoss loss function takes this into account.
        # It essentially computes the softmax first and then computes the negative log-likelihood for the target classes. 
        return logits        

In [470]:
model = SrlModel().to('cuda') # create new model and store weights in GPU memory

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertModel: ['cls.predictions.transform.dense.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.bias', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight', 'cls.predictions.bias']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


Now we are ready to try running the model with just a single input example to check if it is working correctly. Clearly it has not been trained, so the output is not what we expect. But we can see what the loss looks like for an initial sanity check. 

**TODO**: 
* Take a single data item from the dev set, as provided by your Dataset class defined above. Obtain the input token ids, attention mask, predicate indicator mask, and target labels. 
* Run the model on the ids, attention mask, and predicate mask like this: 

In [471]:
# pick an item from the dataset. Then run 
idx = 0
data_item = data[idx]

ids = data_item['ids']
mask = data_item['mask']
pred = data_item['pred']
target_labels = data_item['targets']

ids = ids.to('cuda')
mask = mask.to('cuda')
pred = pred.to('cuda')

outputs = model(ids, mask, pred)
print("Shape of outputs:", outputs.shape)


Shape of outputs: torch.Size([1, 128, 53])


**TODO**: 
Compute the loss on this one item only.
The initial loss should be close to -ln(1/num_labels)

Without training we would assume that all labels for each token (including the target label) are equally likely, so the negative log probability for the targets should be approximately $$-\ln(\frac{1}{\text{num_labels}}).$$ This is what the loss function should return on a single example. This is a good sanity check to run for any multi-class prediction problem. 

In [472]:
import math
-math.log(1 / len(role_to_id), math.e)

3.970291913552122

In [473]:
loss_function = CrossEntropyLoss(ignore_index = -100, reduction='mean')

# complete this. Note that you still have to provide a (batch_size, input_pos) 
# tensor for each parameter, where batch_size =1

# outputs = model(ids, mask, pred) 
# loss = loss_function(...)
# loss.item()   #this should be approximately the score from the previous cell

outputs_reshaped = outputs.view(-1, outputs.size(-1))
target_labels = target_labels.view(-1)
target_labels = target_labels.to(outputs_reshaped.device)

# Compute the loss
loss = loss_function(outputs_reshaped, target_labels)

# Print the loss
print("Computed Loss:", loss.item())

Computed Loss: 4.090612888336182


**TODO**: At this point you should also obtain the actual predictions by taking the argmax over each position.
The result should look something like this (values will differ).

```
tensor([[ 1,  4,  4,  4,  4,  4,  5, 29, 29, 29,  4, 28,  6, 32, 32, 32, 32, 32,
         32, 32, 30, 30, 32, 30, 32,  4, 32, 32, 30,  4, 49,  4, 49, 32, 30,  4,
         32,  4, 32, 32,  4,  2,  4,  4, 32,  4, 32, 32, 32, 32, 30, 32, 32, 30,
         32,  4,  4, 49,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  6,  6, 32, 32,
         30, 32, 32, 32, 32, 32, 30, 30, 30, 32, 30, 49, 49, 32, 32, 30,  4,  4,
          4,  4, 29,  4,  4,  4,  4,  4,  4, 32,  4,  4,  4, 32,  4, 30,  4, 32,
         30,  4, 32,  4,  4,  4,  4,  4, 32,  4,  4,  4,  4,  4,  4,  4,  4,  4,
          4,  4]], device='cuda:0')
```

Then use the id_to_role dictionary to decode to actual tokens. 

```
['[CLS]', 'O', 'O', 'O', 'O', 'O', 'B-ARG0', 'I-ARG0', 'I-ARG0', 'I-ARG0', 'O', 'B-V', 'B-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG1', 'I-ARG2', 'I-ARG1', 'I-ARG2', 'O', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'O', 'I-ARGM-TMP', 'O', 'I-ARGM-TMP', 'I-ARG2', 'I-ARG1', 'O', 'I-ARG2', 'O', 'I-ARG2', 'I-ARG2', 'O', '[SEP]', 'O', 'O', 'I-ARG2', 'O', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG2', 'O', 'O', 'I-ARGM-TMP', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'B-ARG1', 'B-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG1', 'I-ARG1', 'I-ARG2', 'I-ARG1', 'I-ARGM-TMP', 'I-ARGM-TMP', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'O', 'O', 'O', 'O', 'I-ARG0', 'O', 'O', 'O', 'O', 'O', 'O', 'I-ARG2', 'O', 'O', 'O', 'I-ARG2', 'O', 'I-ARG1', 'O', 'I-ARG2', 'I-ARG1', 'O', 'I-ARG2', 'O', 'O', 'O', 'O', 'O', 'I-ARG2', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O']
```

For now, just make sure you understand how to do this for a single example. Later, you will write a more formal function to do this once we have trained the model. 

In [474]:
# Use torch.argmax() to find the index of the maximum value along the predicted logits
print("Shape of outputs:", outputs.shape)

predicted_indices = torch.argmax(outputs, dim=2)

# Print the predicted indices
print(predicted_indices)

# Convert predicted indices to role labels using id_to_role dictionary
predicted_labels = [id_to_role.get(int(idx),id_to_role[4] ) for idx in predicted_indices.squeeze()]

# Print the predicted labels
print(predicted_labels)

Shape of outputs: torch.Size([1, 128, 53])
tensor([[52, 38, 38, 52, 52, 45, 50, 50,  1,  1, 50, 50, 12, 12, 38, 38, 50, 50,
         50, 50, 50, 50, 50, 50, 50, 50, 33, 50, 45, 50, 45, 50, 50, 50,  5, 50,
          1, 11, 11, 25, 23, 23, 25, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
         26, 50, 23, 23, 50, 50, 45, 50,  5, 50, 34, 34, 25, 23, 23, 35, 50, 50,
         50, 50,  4, 50, 20, 50, 50,  4, 50, 50, 50, 45, 29, 50, 33, 45, 45, 45,
         50,  5, 50, 11, 34, 25, 23, 23, 50, 50, 50, 50, 50, 50,  4, 50, 50, 50,
          4, 35, 50, 50, 50, 33, 50, 26, 20, 33, 50, 20,  5, 50, 35, 35, 29,  4,
         25, 50]], device='cuda:0')
['I-V', 'I-ARGM-DIR', 'I-ARGM-DIR', 'I-V', 'I-V', 'I-ARGM-NEG', 'I-ARGM-CXN', 'I-ARGM-CXN', '[CLS]', '[CLS]', 'I-ARGM-CXN', 'I-ARGM-CXN', 'B-ARGM-ADV', 'B-ARGM-ADV', 'I-ARGM-DIR', 'I-ARGM-DIR', 'I-ARGM-CXN', 'I-ARGM-CXN', 'I-ARGM-CXN', 'I-ARGM-CXN', 'I-ARGM-CXN', 'I-ARGM-CXN', 'I-ARGM-CXN', 'I-ARGM-CXN', 'I-ARGM-CXN', 'I-ARGM-CXN', 'I-ARG3', 'I-ARGM-CX

## 3. Training loop

pytorch provides a DataLoader class that can be wrapped around a Dataset to easily use the dataset for training. The DataLoader allows us to easily adjust the batch size and shuffle the data. 

In [475]:
from torch.utils.data import DataLoader
loader = DataLoader(data, batch_size = 32, shuffle = True)

The following cell contains the main training loop. The code should work as written and report the loss after each batch,
cumulative average loss after each 100 batches, and print out the final average loss after the epoch. 

**TODO**: Modify the training loop belowso that it also computes the accuracy for each batch and reports the 
average accuracy after the epoch. 
The accuracy is the number of correctly predicted token labels out of the number of total predictions. 
Make sure you exclude [PAD] tokens, i.e. tokens for which the target label is -100. It's okay to include [CLS] and [SEP] in the accuracy calculation. 

In [476]:
loss_function = CrossEntropyLoss(ignore_index = -100, reduction='mean')

LEARNING_RATE = 1e-05
optimizer = torch.optim.AdamW(params=model.parameters(), lr=LEARNING_RATE)

device = 'cuda'

def train():
    """
    Train the model for one epoch.
    """
    tr_loss = 0 
    nb_tr_examples, nb_tr_steps = 0, 0
    tr_preds, tr_labels = [], []
    tr_correct = 0
    total_tokens = 0
    # put model in training mode
    model.train()
    
    for idx, batch in enumerate(loader):
        
        # Get the encoded data for this batch and push it to the GPU
        ids = batch['ids'].squeeze(1).to(device, dtype=torch.long)
        mask = batch['mask'].squeeze(1).to(device, dtype=torch.long)
        pred_mask = batch['pred'].squeeze(1).to(device, dtype=torch.long)
        targets = batch['targets'].squeeze(1).to(device, dtype=torch.long)
        
        # Run the forward pass of the model
        logits = model(input_ids=ids, attn_mask=mask, pred_indicator=pred_mask)        
        loss = loss_function(logits.transpose(2,1), targets) 
        tr_loss += loss.item()
#         print("Batch loss: ", loss.item()) # can comment out if too verbose.
        
        nb_tr_steps += 1
        nb_tr_examples += targets.size(0)
        
        if idx % 100==0:
            #torch.cuda.empty_cache() # can help if you run into memory issues
            curr_avg_loss = tr_loss/nb_tr_steps
            print(f"Current average loss: {curr_avg_loss}")
                   
        # Compute accuracy for this batch
        matching = torch.sum(torch.argmax(logits,dim=2) == targets)       
        predictions = torch.sum(torch.where(targets==-100,0,1))
        tr_correct += matching.item()
        total_tokens += predictions.item()
                
        # Run the backward pass to update parameters
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    epoch_accuracy = tr_correct/total_tokens
    epoch_loss = tr_loss / nb_tr_steps
    print(f"Training loss epoch: {epoch_loss}")
    print(f"Training accuracy : {epoch_accuracy}")

Now let's train the model for one epoch. This will take a while (up to a few hours). 

In [477]:
train()

Current average loss: 4.0110626220703125
Current average loss: 2.145240839165036
Current average loss: 1.7726128555055876
Current average loss: 1.5824990716091423
Current average loss: 1.4613406054694158
Current average loss: 1.3744974625324775
Current average loss: 1.3072105344639047
Current average loss: 1.2511726419697815
Current average loss: 1.2009285497159399
Current average loss: 1.1553152764213999
Current average loss: 1.1138433446655502
Current average loss: 1.07407389512179
Current average loss: 1.0377615558962143
Current average loss: 1.005381319932439
Current average loss: 0.9757804775944273
Current average loss: 0.9464494012380583
Current average loss: 0.9203385586369268
Current average loss: 0.8967349390504221
Current average loss: 0.8742173810027692
Current average loss: 0.8534381507667098
Current average loss: 0.8346365096121058
Current average loss: 0.8153468017235419
Current average loss: 0.7992839364911253
Current average loss: 0.7831220185624574
Current average loss

In my experiments, I found that two epochs are needed for good performance. 

In [479]:
train() 

Current average loss: 0.12781822681427002
Current average loss: 0.2029448686407344
Current average loss: 0.20881534375213273
Current average loss: 0.21610954948437966
Current average loss: 0.21297501610067124
Current average loss: 0.21108067137544265
Current average loss: 0.20925477927491787
Current average loss: 0.20959725081601768
Current average loss: 0.21055817556161707
Current average loss: 0.21027279801906545
Current average loss: 0.21071981186499367
Current average loss: 0.20993831497934273
Current average loss: 0.2112272777140984
Current average loss: 0.21083716073751357
Current average loss: 0.20959275844457062
Current average loss: 0.20997436514880005
Current average loss: 0.20933877498953882
Current average loss: 0.2097782307402447
Current average loss: 0.20907362910822191
Current average loss: 0.20896443335838158
Current average loss: 0.20887097199519594
Current average loss: 0.20871443501977283
Current average loss: 0.2084377192008203
Current average loss: 0.20911628215604

I ended up with a training loss of about 0.19 and a training accuracy of 0.94. Specific values may differ. 

At this point, it's a good idea to save the model (or rather the parameter dictionary) so you can continue evaluating the model without having to retrain. 

In [480]:
# torch.save(model.state_dict(), "srl_model_fulltrain_2epoch_finetune_1e-05.pt")
torch.save(model.state_dict(), "srl_model_full_train.pt")

## 4. Decoding

In [483]:
# Optional step: If you stopped working after part 3, first load the trained model 

model = SrlModel().to('cuda') 
model.load_state_dict(torch.load("srl_model_full_train.pt"))
model = model.to('cuda')

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertModel: ['cls.predictions.transform.dense.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.bias', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight', 'cls.predictions.bias']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


**TODO (this is the fun part)**: Now that we have a trained model, let's try labeling an unseen example sentence. Complete the functions decode_output and label_sentence below. decode_output takes the logits returned by the model, extracts the argmax to obtain the label predictions for each token, and then translate the result into a list of string labels. 

label_sentence takes a list of input tokens and a predicate index, prepares the model input, call the model and then call decode_output to produce a final result. 

Note that you have already implemented all components necessary (preparing the input data from the token list and predicate index, decoding the model output). But now you are putting it together in one convenient function. 

In [484]:
tokens = "A U. N. team spent an hour inside the hospital , where it found evident signs of shelling and gunfire .".split()

In [485]:
def decode_output(logits): # it will be useful to have this in a separate function later on
    """
    Given the model output, return a list of string labels for each token. 
    """
    # Extract the predicted labels for each token
    predicted_labels = torch.argmax(logits, dim=2)
#     print("predicted_labels")
#     print(predicted_labels)
    # Translate the predicted label indices into string labels
#     labels = [id_to_role[label_id.item()] for label_id in predicted_labels[0]]
    labels = [id_to_role.get(int(label_id), id_to_role[4])  for label_id in predicted_labels[0]]
#     label_ids = [role_to_id.get(label, role_to_id['O']) for label in token_labels]

    
    return labels
    

In [486]:
def label_sentence(tokens, pred_idx):
    """
    Given a list of input tokens and a predicate index, label the sentence.
    """
    # complete this function to prepare token_ids, attention mask, predicate mask, then call the model. 
    # Decode the output to produce a list of labels. 
    token_ids = tokenizer.convert_tokens_to_ids(tokens)

    attention_mask = [1 if token != -100 else 0 for token in token_ids]
    pred_indicator_mask = [1 if i == pred_idx else 0 for i in range(len(tokens))]
    token_tensor = torch.tensor([token_ids], dtype=torch.long).to(device)
    mask_tensor = torch.tensor([attention_mask], dtype=torch.long).to(device)
    pred_tensor = torch.tensor([pred_indicator_mask], dtype=torch.long).to(device)
    
    with torch.no_grad():
        logits = model(token_tensor, mask_tensor, pred_tensor)
    return decode_output(logits)
        
    

In [487]:
# Now you should be able to run

label_test = label_sentence(tokens, 13) # Predicate is "found"
zip(tokens, label_test)
# Assuming tokens is a list of tokens and label_test is a list of corresponding labels
for token, label in zip(tokens, label_test):
    print((token, label))


('A', 'O')
('U.', 'O')
('N.', 'O')
('team', 'O')
('spent', 'O')
('an', 'O')
('hour', 'O')
('inside', 'O')
('the', 'O')
('hospital', 'O')
(',', 'O')
('where', 'O')
('it', 'O')
('found', 'O')
('evident', 'O')
('signs', 'O')
('of', 'O')
('shelling', 'O')
('and', 'O')
('gunfire', 'O')
('.', 'O')


The expected output is somethign like this: 
```   
 ('A', 'O'),
 ('U.', 'O'),
 ('N.', 'O'),
 ('team', 'O'),
 ('spent', 'O'),
 ('an', 'O'),
 ('hour', 'O'),
 ('inside', 'O'),
 ('the', 'B-ARGM-LOC'),
 ('hospital', 'I-ARGM-LOC'),
 (',', 'O'),
 ('where', 'B-ARGM-LOC'),
 ('it', 'B-ARG0'),
 ('found', 'B-V'),
 ('evident', 'B-ARG1'),
 ('signs', 'I-ARG1'),
 ('of', 'I-ARG1'),
 ('shelling', 'I-ARG1'),
 ('and', 'I-ARG1'),
 ('gunfire', 'I-ARG1'),
 ('.', 'O'),
```


### 5. Evaluation 1: Token-Based Accuracy
We want to evaluate the model on the dev or test set. 

In [488]:
# Open the input file
with open('propbank_dev.tsv', 'r') as input_file:
    # Read the first 10 lines
    lines = input_file.readlines()[:1600]

# Open the output file
with open('output_dev.tsv', 'w') as output_file:
    # Write the lines to the output file
    output_file.writelines(lines)

In [489]:
dev_data = SrlData("output_dev.tsv") # Takes a while because we preprocess all data offline

In [490]:
from torch.utils.data import DataLoader
loader = DataLoader(dev_data, batch_size = 1, shuffle = False)

In [491]:
# # Optional: Load the model again if you stopped working prior to this step. 
# model = SrlModel()
# model.load_state_dict(torch.load("srl_model_8000_train.pt"))
# model = mode.to('cuda')

**TODO**: Complete the evaluate_token_accuracy function below. The function should iterate through the items in the data loader (see training loop in part 3). Run the model on each sentence/predicate pair and extract the predictions.

For each sentence, count the correct predictions and the total predictions. Finally, compute the accuracy as #correct_predictions / #total_predictions

Careful: You need to filter out the padded positions ([PAD] target tokens), as well as [CLS] and [SEP]. It's okay to include [B-V] in the count though. 

In [494]:
def evaluate_token_accuracy(model, loader):
    
    model.eval() # put model in evaluation mode
    
    # for the accuracy 
    tr_loss = 0 
    nb_tr_examples, nb_tr_steps = 0, 0
    tr_preds, tr_labels = [], []
    tr_correct = 0
    total_tokens = 0
    
    # iterate over the data here. 
    for idx, batch in enumerate(loader):
        
        # Get the encoded data for this batch and push it to the GPU
        ids = batch['ids'].squeeze(1).to(device, dtype=torch.long)
        mask = batch['mask'].squeeze(1).to(device, dtype=torch.long)
        pred_mask = batch['pred'].squeeze(1).to(device, dtype=torch.long)
        targets = batch['targets'].squeeze(1).to(device, dtype=torch.long)
        
        # Run the forward pass of the model
        logits = model(input_ids=ids, attn_mask=mask, pred_indicator=pred_mask)        
        loss = loss_function(logits.transpose(2,1), targets) 
        # Compute accuracy for this batch
#         print(torch.argmax(logits,dim=2))
        matching = torch.sum(torch.argmax(logits,dim=2) == targets)       
        predictions = torch.sum(torch.where(targets==-100,0,1))
        tr_correct += matching.item()
        total_tokens += predictions.item()
        
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    epoch_accuracy = tr_correct/total_tokens
    print(f"accuracy : {epoch_accuracy}")

#     acc = total_correct / total_predictions
#     print(f"Accuracy: {acc}")
    

In [495]:
evaluate_token_accuracy(model, loader)

accuracy : 0.9287232965475347


### 6. Span-Based evaluation 

While the accuracy score in part 5 is encouraging, an accuracy-based evaluation is problematic for two reasons. First, most of the target labels are actually O. Second, it only tells us that per-token prediction works, but does not directly evaluate the SRL performance. 

Instead, SRL systems are typically evaluated on micro-averaged precision, recall, and F1-score for predicting labeled spans. 

More specifically, for each sentence/predicate input, we run the model, decode the output, and extract a set of labeled spans (from the output and the target labels). These spans are (i,j,label) tuples.  

We then compute the true_positives, false_positives, and false_negatives based on these spans. 

In the end, we can compute 

* Precision:  true_positive / (true_positives + false_positives)  , that is the number of correct spans out of all predicted spans. 

* Recall: true_positives / (true_positives + false_negatives) , that is the number of correct spans out of all target spans. 

* F1-score:   (2 * precision * recall) / (precision + recall)


For example, consider 

| |[CLS]|The|judge|scheduled|to|preside|over|his|trial|was|removed|from|the|case|today|.|             
|--||---|-----|---------|--|-------|----|---|-----|---|-------|----|---|----|-----|-|             
||0|1|2|3|4|5|6|7|8|9|1O|11|12|13|14|15|
|target|[CLS]|B-ARG1|I-ARG1|B-V|B-ARG2|I-ARG2|I-ARG2|I-ARG2|I-ARG2|O|O|O|O|O|O|O|
|prediction|[CLS]|B-ARG1|I-ARG1|B-V|I-ARG2|I-ARG2|O|O|O|O|O|O|O|O|B-ARGM-TMP|O|

The target spans are (1,2,"ARG1"), and (4,8,"ARG2").

The predicted spans would be (1,2,"ARG1"), (14,14,"ARGM-TMP"). Note that in the prediction, there is no proper ARG2 span because we are missing the B-ARG2 token, so this span should not be created. 

So for this sentence we woudl get: true_positives: 1 false_positives: 1 false_negatives: 1

*TODO*: Complete the function evaluate_spans that performs the span-based evaluation on the given model and data loader. You can use the provided extract_spans function, which returns the spans as a dictionary. For example
{(1,2): "ARG1", (4,8):"ARG2"} 

In [496]:
def extract_spans(labels):
    spans = {} # map (start,end) ids to label
    current_span_start = 0
    current_span_type = ""
    inside = False
    for i, label in enumerate(labels):
        if label.startswith("B"):            
            if inside: 
                if current_span_type != "V":
                    spans[(current_span_start,i)] = current_span_type            
            current_span_start = i
            current_span_type = label[2:]
            inside = True
        elif inside and label.startswith("O"):
            if current_span_type != "V":
                spans[(current_span_start,i)] = current_span_type
            inside = False
        elif inside and label.startswith("I") and label[2:] != current_span_type:            
            if current_span_type != "V":
                spans[(current_span_start,i)] = current_span_type
            inside = False
    return spans
                        

In [497]:
def evaluate_spans(model, loader):
    
    total_tp = 0
    total_fp = 0
    total_fn = 0
        
    for idx, batch in enumerate(loader):
        # Get the encoded data for this batch and push it to the GPU
        ids = batch['ids'].squeeze(1).to(device, dtype=torch.long)
        mask = batch['mask'].squeeze(1).to(device, dtype=torch.long)
        pred_mask = batch['pred'].squeeze(1).to(device, dtype=torch.long)
        targets = batch['targets'].squeeze(1).to(device, dtype=torch.long)
        
        # Run the forward pass of the model
        logits = model(input_ids=ids, attn_mask=mask, pred_indicator=pred_mask)
        predictions = torch.argmax(logits, dim=2)
        predicted_labels = [id_to_role.get(int(idx), id_to_role[4])for idx in predictions.squeeze()]
        target_labels = [id_to_role.get(int(idx), id_to_role[4])for idx in targets.squeeze()]
        
        # Decode spans from predictions and targets
        pred_spans = extract_spans(predicted_labels)
        target_spans = extract_spans(target_labels)
            # Calculate true positives, false positives, and false negatives
        for span in pred_spans:
            if span in target_spans:
                total_tp += 1
            else:
                total_fp += 1
        for span in target_spans:
            if span not in pred_spans:
                total_fn += 1
                    
    total_p = total_tp / (total_tp + total_fp)
    total_r = total_tp / (total_tp + total_fn)
    total_f = (2 * total_p * total_r) / (total_p + total_r)
            
    print(f"Overall P: {total_p}  Overall R: {total_r}  Overall F1: {total_f}")

# Call the function
evaluate_spans(model, loader)

Overall P: 0.15443548387096775  Overall R: 0.8855491329479769  Overall F1: 0.26300429184549357


In my evaluation, I got an F score of 0.82  (which slightly below the state-of-the art in 2018)

### OPTIONAL: 

Repeat the span-based evaluation, but print out precision/recall/f1-score for each role separately.